arroyo/processing/strategies/produce.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import time
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, Optional, Tuple
+
+from arroyo.backends.abstract import Producer
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import Commit, Message, Position, Topic, TPayload
+
+logger = logging.getLogger(__name__)
+
+
+class ProduceAndCommit(ProcessingStrategy[TPayload]):
+    """
+    For each message received, produces a message to the topic being passed in.
+    Waits on the futures and only once completed, commits the offset of the message.
+    """
+
+    def __init__(
+        self,
+        producer: Producer[TPayload],
+        topic: Topic,
+        commit: Commit,
+        max_buffer_size: int = 10000,
+    ):
+        self.__producer = producer
+        self.__topic = topic
+        self.__commit = commit
+        self.__max_buffer_size = max_buffer_size
+
+        self.__queue: Deque[
+            Tuple[Message[TPayload], Future[Message[TPayload]]]
+        ] = deque()
+
+        self.__closed = False
+
+    def poll(self) -> None:
+        while self.__queue:
+            message, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            self.__queue.popleft()
+
+            self.__commit(
+                {message.partition: Position(message.next_offset, message.timestamp)}
+            )
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+
+        if len(self.__queue) >= self.__max_buffer_size:
+            raise MessageRejected
+
+        self.__queue.append(
+            (message, self.__producer.produce(self.__topic, message.payload))
+        )
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+
+        # Commit all pending offsets
+        self.__commit({}, force=True)
```

Comment:
I think you should commit after having waited for the callbacks. You can still receive errors from them.

Reply:
The comment may be misleading, I will update it. This is only committing already staged and succesfully produced offsets.